examples/vision/aodnet.py

```diff
@@ -0,0 +1,376 @@
+# Downloading Dataset
+
+The authors of the AODNet paper have created synthesized hazy images using the
+groundtruth images with depth meta-data from the indoor [NYU2
+Depth Database](https://cs.nyu.edu/~silberman/datasets/nyu_depth_v2.html). We will
+download this dataset from Google Drive using `gdown`.
+"""
+
+gdown.download(
+    "https://drive.google.com/uc?id=1sInD9Ydq8-x7WwqehE0EyRknMdSFPOat",
+    "Dehaze-NYU.zip",
+    quiet=False,
+)
+
+os.system("unzip -qq Dehaze-NYU.zip")
+os.system("rm ./Dehaze-NYU.zip")
+
+"""
+## Data Loader
+
+In order to ensure that the model is fed the data efficiently during training, we will be
+use the `tf.data` API to create our data loading pipeline. Due to memory constraints for
+training on Google Colab, we generate random crops of size `256` from corresponding
+hazing and clean image pair. For the training dataset, we use also apply random rotation
+and random horizontal flips in order to augment the data.
+"""
+
+
+def get_image_file_list(dataset_path):
+    original_image_files = []
+    hazy_image_paths = sorted(
+        glob(str(os.path.join(dataset_path, "train_images/*.jpg")))
+    )
+    for image_path in hazy_image_paths:
+        image_file_name = image_path.split("/")[-1]
+        original_file_name = (
+            image_file_name.split("_")[0] + "_" + image_file_name.split("_")[1] + ".jpg"
+        )
+        original_image_files.append(
+            str(os.path.join(dataset_path, "original_images/" + original_file_name))
+        )
+    return original_image_files, hazy_image_paths
+
+
+def read_images(image_files):
+    dataset = tf.data.Dataset.from_tensor_slices(image_files)
+    dataset = dataset.map(tf.io.read_file)
+    dataset = dataset.map(
+        lambda x: tf.image.decode_png(x, channels=3),
+        num_parallel_calls=tf.data.experimental.AUTOTUNE,
+    )
+    return dataset
+
+
+def random_crop(hazy_image, original_image, low_crop_size, enhanced_crop_size):
+    hazy_image_shape = tf.shape(hazy_image)[:2]
+    low_w = tf.random.uniform(
+        shape=(), dtype=tf.int32, maxval=hazy_image_shape[1] - low_crop_size + 1
+    )
+    low_h = tf.random.uniform(
+        shape=(), dtype=tf.int32, maxval=hazy_image_shape[0] - low_crop_size + 1
+    )
+    enhanced_w = low_w
+    enhanced_h = low_h
+    hazy_image_cropped = hazy_image[
+        low_h : low_h + low_crop_size, low_w : low_w + low_crop_size
+    ]
+    original_image_cropped = original_image[
+        enhanced_h : enhanced_h + enhanced_crop_size,
+        enhanced_w : enhanced_w + enhanced_crop_size,
+    ]
+    return hazy_image_cropped, original_image_cropped
+
+
+def random_flip(hazy_image, original_image):
+    return tf.cond(
+        tf.random.uniform(shape=(), maxval=1) < 0.5,
+        lambda: (hazy_image, original_image),
+        lambda: (
+            tf.image.flip_left_right(hazy_image),
+            tf.image.flip_left_right(original_image),
+        ),
+    )
+
+
+def random_rotate(hazy_image, original_image):
```

Comment:
Random rotation by a factor of 90 degrees aren't useful to augment natural images (since they have a standard orientation). Instead, we should prefer random rotations by a small angle. You can do this with the `RandomRotation` preprocessing layer.